src/world.rs

```diff
@@ -0,0 +1,150 @@
+use std::{
+    io::Read,
+    path::{Path, PathBuf},
+};
+
+use regex::Regex;
+use serde::Deserialize;
+
+use crate::{Error, ResourceReader};
+
+/// A World is a list of maps files or regex patterns that define a layout of TMX maps.
+/// You can use the loader to further load the maps defined by the world.
+#[derive(Deserialize, PartialEq, Clone, Debug)]
+pub struct World {
+    /// The path first used in a [`ResourceReader`] to load this world.
+    #[serde(skip_deserializing)]
+    pub source: PathBuf,
+    /// The [`WorldMap`]s defined in the world file.
+    pub maps: Option<Vec<WorldMap>>,
+    /// Optional regex pattern to load maps.
+    pub patterns: Option<Vec<WorldPattern>>,
+    /// The type of world, which is arbitrary and set by the user.
+    #[serde(rename = "type")]
+    pub world_type: Option<String>,
+}
+
+/// A WorldMap provides the information for a map in the world and its layout.
+#[derive(Deserialize, PartialEq, Clone, Debug)]
+pub struct WorldMap {
+    /// The filename of the tmx map.
+    #[serde(rename = "fileName")]
+    pub filename: String,
+    /// The x position of the map.
+    pub x: i32,
+    /// The y position of the map.
+    pub y: i32,
+    /// The optional width of the map.
+    pub width: Option<u32>,
+    /// The optional height of the map.
+    pub height: Option<u32>,
+}
+
+/// A WorldPattern defines a regex pattern to automatically determine which maps to load and how to lay them out.
+#[derive(Deserialize, PartialEq, Clone, Debug)]
+pub struct WorldPattern {
+    /// The regex pattern to match against filenames. The first two capture groups should be the x integer and y integer positions.
+    pub regexp: String,
+    /// The multiplier for the x position.
+    #[serde(rename = "multiplierX")]
+    pub multiplier_x: u32,
+    /// The multiplier for the y position.
+    #[serde(rename = "multiplierY")]
+    pub multiplier_y: u32,
+    /// The offset for the x position.
+    #[serde(rename = "offsetX")]
+    pub offset_x: i32,
+    /// The offset for the y position.
+    #[serde(rename = "offsetY")]
+    pub offset_y: i32,
+}
+
+impl WorldPattern {
+    /// Utility function to test a single path against the defined regexp field and returns a parsed WorldMap if it matches.
+    /// Returns none if the filename does not match the pattern.
+    pub fn capture_path(&self, path: &Path) -> Result<WorldMap, Error> {
+        let re = Regex::new(&self.regexp).unwrap();
```

Comment:
Since there's a cost involved in parsing the regex, it would be nice if `self.regexp` was already a `Regex` instead of a `String`. The `.unwrap()` should probably be changed to cause a world loading error.

Reply:
I added the serde_regex crate to the `world` feature. I tried to avoid using it but the de-serialization rabbit hole would get to the point where most of the world.rs file would be dealing with that single variable.